### ISSUE TITLE
Hypervisor

### ISSUE THREAD

**Issue by danxn** [runtimejs/runtime#42]
Hello, sorry for a bit stupid question, I just started to learn runtime.js.

My question is - is it difficult to write hypervisor with runtime.js?
All, what I need is convinient "wrapper" for distributed virtualization system and tool which will allow to create in comparatively easy way RDP or VNC-like protocol.

I know that cloud systems is very popular now on server-side, but my idea is to connect mobile devices like smartphones, tablets and laptops in one personal cloud, or personal computing network, which will allow to run different OSes and applications. The main goal - is to combine computing performance of each device and make one computer from all this stuff with help of network GUI access to each of them from one terminal.

It is also can be usefull for server-side javascript programming, of course, but first I interested in personal application - when each new (actually - old) device like smartphone or tabled can be connected to this personal cloud and increase it calculating performance. 

I've experimented a bit with Android LinuxDeploy, VNC and few smartphones and tablets - it looks nice. Each VNC session looks just like additional dashboard and all data can be shared via network disks. But it looks a bit like frankenstain, and may be runtime.js can provide suitable tool for creating this system with all it's components on low level with relatively small efforts?

**Comment by iefserge** [comment]
@danxn Hello, thank you for question. Sorry for the missing or incomplete runtime.js documentation, I hope I'll be able to update it soon, as well as upload new website.

Current runtime.js version needs KVM host to run on top of (runtime.js as a guest machine). This limitation is mainly because there is no support (aka drivers) for physical hardware and the system uses [virtio](http://wiki.libvirt.org/page/Virtio) as the only way to communicate with the outside world. The only supported architecture is x86_64 (though it should be relatively easy to create i386 port. are there 64 bit smartphones?).

I don't really know how Android LinuxDeploy works. Does it run a VM on Android? If that's the case it might be possible to also run KVM and runtime.js there.

**Comment by danxn** [comment]
@iefserge Hi, thanks for explanation it makes whole picture more clear.

>Current runtime.js version needs KVM host to run on top of (runtime.js as a guest machine). 
>This limitation is mainly because there is no support (aka drivers) for physical hardware 
>and the system uses virtio as the only way to communicate with the outside world.

So, the runtime.js for now is far from turning into hypervisor and works inside KVM hypervisor? Ok, not a problem, anyway possibility to write and run compiled JS for servers is a big advantage. Accounting on it's relative simplicity - it's really cool platform for web server in cloud.

> The only supported architecture is x86_64 (though it should be relatively easy to create i386 port. are there 64 bit smartphones?).

Yes, i've made a mistake - Intel Atom SoCs supports x86_64 architecture, but not in all series:
[Atom_(system_on_chip)](http://en.wikipedia.org/wiki/Atom_(system_on_chip)),

As [here](http://en.wikipedia.org/wiki/Intel_Atom_(CPU)) said: *All Atom processors implement the x86 (IA-32) instruction set; however, support for the Intel 64 instruction set was not added until the desktop Diamondville and desktop and mobile Pineview cores. _The Atom N2xx and Z5xx series Atom models cannot run x86-64 code._[21] The Centerton server processors will support the Intel 64 instruction set.[7]*

Mostly spreaded Atom SoCs for smartphones should support 64-bit instructions set (it's mainly Z2xx and Z3xx series): [quick price search](http://hotline.ua/mobile/mobilnye-telefony-i-smartfony/66683/). The choice of tablets is wider: [price search results](http://hotline.ua/computer/planshety/37791-21393/)

>I don't really know how Android LinuxDeploy works. Does it run a VM on Android? If that's the case it might be possible to also run KVM and runtime.js there.

It's open source tool, you can find it here - [LinuxDeploy repo](https://github.com/meefik/linuxdeploy). [Here](http://www.linux-magazine.com/Online/Features/Convert-an-Android-Device-to-Linux) the article which describes in details with screenshoots how it is used (chapter "Going the chroot Root: Linux Deploy"). Linux Deploy uses [chroot](http://en.wikipedia.org/wiki/Chroot#Uses) to run on the same machine where Android runs, they are working together at the same time on the same hardware. Works fine, I've tested it on ARM smartphone and ARM/x86-64(Atom Z3xx) tablets - nodejs and express works just fine, but for ARM node should be compiled from sources. So the Linux KVM with runtime.js should work as well.

I'm not sure, but may be mobile SoCs is a good direction to move runtime.js from virtual host to hardware envieronment? I just don't know what drivers set needed for SoCs, but theoretically it should be more stable than for standalone CPU.

**Comment by iefserge** [comment]
@danxn 
> So, the runtime.js for now is far from turning into hypervisor

Yep, the initial idea is to be able to create small, immutable and specialized runtime.js VM images that include application code and only necessary system components. Those images then could be deployed into cloud.

> I'm not sure, but may be mobile SoCs is a good direction to move runtime.js from virtual host to hardware envieronment?

It would be very interesting to run it on mobile SoC, if anyone interested wants to hack on this. I'm curious, what's the use of such a runtime.js powered mobile device?

**Comment by piranna** [comment]
> It would be very interesting to run it on mobile SoC, if anyone interested wants to hack on this. I'm curious, what's the use of such a runtime.js powered mobile device?

Raspberry Pi, do i need to say more? :-D

**Comment by danxn** [comment]
@iefserge 
> It would be very interesting to run it on mobile SoC, if anyone interested wants to hack on this. I'm curious, what's the use of such a runtime.js powered mobile device?

Well, the performance of contemporary mobile SoCs is pretty close to performance of 5-years-old top server platforms. And more than enough if you need something like web-shop or blog. Especially if this SoCs present in more than one instance and runs cloud. Accounting on their size and low requirements to cooling and power system this is very intresting alternative for collocation, for example. My goal I've described in first post - I need "modular" PC which will be assembled from "cubicles" like smartphones and tablets (and other stuff), each element will increase whole system performance almost in linear manner and resources distribution will be done on applications level. Pretty cool computer, I think :)

@piranna 
> Raspberry Pi, do i need to say more? :-D

Yes, sure, and not only Raspberry PI. There is a lot of minicomputers based on mobile SoCs - like Intel Galileo, pcDuino or Cubiboard. There is also exist some projects which runs JS on very low-level SoCs, like Arduino. Like [Espruino](http://www.espruino.com/) and [Onion Omega](https://www.kickstarter.com/projects/onion/onion-omega-invention-platform-for-the-internet-of/description). 

So, in short:
1) Robots
2) CNCs/3D Printers
3) Smart house applications

I.e. Internet of the things. 10MB OS with all V8 functional? Pretty cool for data processing I think.

**Comment by iefserge** [comment]
@danxn @piranna Ah, that's very cool. Looks like runtime.js should be able to boot on Intel 64 bit chips (it works on my Intel Desktop PC), but without any access to network. Driver would be necessary to make this work. Wouldn't say it's hard to write one, but would require some pretty low-level JavaScript code :)

**Comment by piranna** [comment]
Basic keyboard and VGA control is somewhat easy and just enough for a hello world... :-) Don't know how much difficult would be to program a network card, but at least the generic ones availables on QEmu or VmWare should be throroughlty documented.

(And a minimal Node.js compatibility, and we have there NodeOS smashing cloud server everywhere :-P)

**Comment by danxn** [comment]
@iefserge 
> Looks like runtime.js should be able to boot on Intel 64 bit chips (it works on my Intel Desktop PC), but without any access to network.

Ok, great.

Take a look at this - 

![image](https://cloud.githubusercontent.com/assets/12805166/8069908/539bbd20-0f06-11e5-9e3f-d8ab8275c9f0.png)

This is Intel Edison (new version of Intel Galileo).

* Its dimensions are 35.5 x 25 x 3.9 mm, with components on both sides.
* The board's main SoC is a 22 nm Intel Atom "Tangier" (Z34XX) that includes two Atom Silvermont cores running at 500 MHz and one Intel Quark core at 100 MHz (for executing RTOS ViperOS). 
* The SoC has 1 GB RAM integrated on package. 
* There is also:
  * 4 GB eMMC flash on board, 
  * Wi-Fi, 
  * Bluetooth 4 
  * and USB controllers. 
* The board has 70-pin dense connector (Hirose DF40) with USB, SD, UARTs, GPIOs. 
* The price of the device is around 50 USD.[13] 
* It runs Yocto Linux with development support for:
  * Arduino IDE, 
  * Eclipse (C, C++, Python), 
  * Intel XDK (NodeJS, HTML5), 
  * and Wolfram.[14][15]

http://en.wikipedia.org/wiki/Intel_Edison

[https://github.com/tarantool/tarantool/wiki/Intel-Edison](Here /proc/cpuinfo) can be found.

Looks like this is good platform to teach runtime.js "to swim" in hardware :)

How do you think? Looks like it needs only network driver to use WiFi, and it will be possible to run "half-metal node server" and to test it's performance on 500MHz core and compare it with similar software instance (Edison supports node.js on top of the Yokto Linux).

> Driver would be necessary to make this work. Wouldn't say it's hard to write one, but would require some pretty low-level JavaScript code :)

Well, I never wrote drivers for network interface cards, but everything once should be done for the first time :) May be you can give some links for quick intro to this?

**Comment by danxn** [comment]
@piranna
> and we have there NodeOS smashing cloud server everywhere

Do you mean this all what is needed to setup cloud platform for nodejs apps? Wow, it's like a bomb. From my point of view, of course.

**Comment by piranna** [comment]
> @piranna
>>
>> and we have there NodeOS smashing cloud server everywhere
>
> Do you mean this all what is needed to setup cloud platform for nodejs
apps? Wow, it's like a bomb. From my point of view, of course.

Network card (I/O) and maybe a hard disk for data storage, since you
already have CPU and RAM. Do you really think is it needed anymore for an
OS running miles away on a data center? And by using a networked data
storage like Amazon S3 or Dropbox you only need the network card support
:-) How do they do to connect PICs and Arduinos to internet? What's the
Internet of Things more than a lot of dumb devices without any local
interface working as micro web servers? :-D

**Comment by piranna** [comment]
Ok, maybe Node.js support would need more work (mostly API compatibility
things), but it's already Javascript and v8 and the only special thing is
libuv, so... :-)

**Comment by danxn** [comment]
@piranna 

> What's the
> Internet of Things more than a lot of dumb devices without any local
> interface working as micro web servers? :-D

Yes, definitely.

**Comment by iefserge** [comment]
@danxn Here is an example of (unfinished) RTL8139 driver https://github.com/runtimejs/runtime/blob/master/initrd/driver/rtl8139.js. Device is controlled by reading/writing memory mapped registers exposed through ArrayBuffer in JavaScript.

`lspci` on Linux could be used to find out which network controller (vendorId and deviceId codes) the board has. Then it should be pretty easy to find a device spec online.

btw, runtime.js is able to scan PCI devices as well, https://github.com/runtimejs/runtime/blob/libos/js/core/pci/scan.js#L815-L818. Could console.log that output as well :)

**Comment by piranna** [comment]
Awesome :-)

**Comment by danxn** [comment]
@iefserge 
> Then it should be pretty easy to find a device spec online.

Well, this is the problem for Edison, because of BCM43340 (wifi/bluetooth module). It's proprietary part of Edison platform and should be used with Broadcom firmware:

https://github.com/instantinfrastructure/edison-src/tree/master/broadcom_cws/wlan
License: https://github.com/instantinfrastructure/edison-src/blob/master/broadcom_cws/wlan/firmware/LICENCE.broadcom_bcm43xx

Is it possible to use this kind of drivers with runtime.js?

And one more question - as I can see the size (in lines, files and folders) of existing BCM43xx driver is much larger than rtl8139.js Is it possible to make runtime.js driver for bcm43340 a bit smaler? :) Or it should be probably larger than runtime.js repo?

**Comment by piranna** [comment]
Speaking in Node.js terms, the firmware could be added some Javascript bindings and require() it as a compiled module... Don't know how much difficult it could be on runtime.js, though...

**Comment by piranna** [comment]
Anyway, before wireless, I would write an ethernet network driver, they are easier since you don't need to take in account rf.

**Comment by danxn** [comment]
@piranna 
> Speaking in Node.js terms, the firmware could be added some Javascript bindings and require() it as a compiled module... Don't know how much difficult it could be on runtime.js, though...

Well, yes, I think it's not a big problem from programming side, but may be some license limitations or something else exists for runtime.js

**Comment by piranna** [comment]
Quick read I don't find any problem there, re-minds me to MIT license but not sure. Anyway, we don't need to copy the code, just download and compile it on install time... :-) I find it more problematic about how to use it: is it a library, or a kernel module?

**Comment by danxn** [comment]
@piranna 
> Anyway, before wireless, I would write an ethernet network driver, they are easier since you don't need to take in account rf.

This very easy if you use "more opened" modules, like [ESP8266](http://www.instructables.com/id/noobs-guide-to-ESP8266-with-Arduino-Mega-2560-or-U/?ALLSTEPS), for example:
https://gist.github.com/teos0009/2f61b6b031a6eb35c48b

This code even shorter than rtl8139.js But in case with BCM43340 - this looks like building OS from scratch.

@piranna 
> Quick read I don't find any problem there, re-minds me to MIT license but not sure. Anyway, we don't need to copy the code, just download and compile it on install time... :-) I find it more problematic about how to use it: is it a library, or a kernel module?

Good question. May be this will explain something: https://github.com/instantinfrastructure/edison-src/tree/master/device-software/meta-edison/recipes-kernel/bcm43340

**Comment by danxn** [comment]
May be usage of ESP8266 is better solution for now.

**Comment by danxn** [comment]
One more usefull repo: https://github.com/01org/edison-bcm43340

**Comment by danxn** [comment]
@iefserge 

@danxn
> Is it possible to make runtime.js driver for bcm43340 a bit smaler? :)

Ok, I see it will depend on common device controlling schema complicity (i've compared rtl8139.js with this RTL8139 driver - http://lxr.free-electrons.com/source/drivers/net/ethernet/realtek/8139too.c)

**Comment by danxn** [comment]
Holy... 

[PCI Local Bus Specification](http://www.ics.uci.edu/~harris/ics216/pci/PCI_22.pdf)

Ok, I'll disappear from thread for a while :)

**Comment by piranna** [comment]
> PCI Local Bus Specification

Cool! :-)

**Comment by iefserge** [comment]
@danxn runtime.js handles all the PCI stuff and wraps memory/IRQ resources into JS objects. Driver code just needs to ask for a specific device using vendor/device codes
https://github.com/runtimejs/runtime/blob/libos/js/driver/virtio/index.js#L36

**Comment by danxn** [comment]
@iefserge 
>  runtime.js handles all the PCI stuff and wraps memory/IRQ resources into JS objects.

phew.. ok  XD

> Driver code just needs to ask for a specific device using vendor/device codes

I've found this list in BCM43340 driver repo - https://github.com/01org/edison-bcm43340/blob/master/common/include/devctrl_if/wlioctl_defs.h

Looks like this is what is needed?

**Comment by danxn** [comment]
One problem with Edison - it's unclear what exactly device is BCM43340:

**Comment by danxn** [comment]
BTW This is known issue for Intel Edison:

https://communities.intel.com/message/269591

>BCM43340 Programmer's Guide

>Kudzu Dec 21, 2014 9:16 AM

>Unfortunately BC has a long history of not publicly providing SDKs or programmers docs much to the frustration of the Linux, WRT, etc community. It appears BC has opened somewhat with some docs on their site for some chipsets but the BC43340 is nowhere to be found. This seems an odd choice for an open source platform.
 
>Short of reverse engineering the driver in the Yocto build, how can on obtain the software interface docs for the 43340? We are planning to port Cosmos (C# Open Source Managed Operating System) (http://www.goCosmos.org) to the Edison and will need to talk directly to the wifi chipset.

**Comment by iefserge** [comment]
@danxn Ah, couldn't find any specs for BCM43340 either, and its driver code looks very complex. My guess is that it would extremely hard to support it.

**Comment by danxn** [comment]
@iefserge 
> My guess is that it would extremely hard to support it.

Yeah, I've tired to browse available driver sources for this chipset (BCM43340). Ok, nevermind. May be when runtime.js will be proted to ARM it will be easy to find suitable network adapter.

Anyway you got the idea: with help of SoCs it is possible to create hardware platform for appropriate software. Like Raspberry PI, but specialized for your needs. This is not very expensive now, and can be done quick enough. For example you can use Raspberry PI project experience. You see - how many analogs appear today? Manufacturers of this devices can place any SoC you need to the board. And you can support your own hardware platform for micro-cloud services. This very intresting thing :)

**Comment by danxn** [comment]
Guess it's close to you - [IKT](http://ictech.com.ua/service/contract-electronics-manufacturing.html) you can ask them for prices. From my experience prototype will cost about 1000-3000, plus small series something around 5000-10000$

**Comment by danxn** [comment]
Or you can try something like barebone-nettop like Intel NUC.
